I see this error in the CI log:
```
2023-10-12T13:08:18.6284364Z [RUSTC-TIMING] libffi_sys test:false 0.107
2023-10-12T13:08:18.6284660Z [RUSTC-TIMING] libffi test:false 0.729
2023-10-12T13:08:18.6285313Z [0m[1m[38;5;9merror[E0432][0m[0m[1m: unresolved import `std::sync::atomic::AtomicU64`[0m
2023-10-12T13:08:18.6285702Z [0m [0m[0m[1m[38;5;12m--> [0m[0msrc/tools/miri/src/clock.rs:1:25[0m
2023-10-12T13:08:18.6285992Z [0m  [0m[0m[1m[38;5;12m|[0m
2023-10-12T13:08:18.6286407Z [0m[1m[38;5;12m1[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0muse std::sync::atomic::{AtomicU64, Ordering};[0m
2023-10-12T13:08:18.6287111Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9m^^^^^^^^^[0m
2023-10-12T13:08:18.6287524Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9m|[0m
2023-10-12T13:08:18.6287971Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9mno `AtomicU64` in `sync::atomic`[0m
2023-10-12T13:08:18.6288420Z [0m  [0m[0m[1m[38;5;12m| [0m[0m                        [0m[0m[1m[38;5;9mhelp: a similar name exists in the module: `AtomicU32`[0m
2023-10-12T13:08:18.6288748Z [0m  [0m[0m[1m[38;5;12m|[0m
2023-10-12T13:08:18.6289224Z [0m  [0m[0m[1m[38;5;12m= [0m[0m[1mhelp[0m[0m: consider importing this type alias instead:[0m
2023-10-12T13:08:18.6289637Z [0m          rustc_data_structures::sync::AtomicU64[0m
2023-10-12T13:08:18.6289801Z 
2023-10-12T13:08:18.6290023Z [0m[1mFor more information about this error, try `rustc --explain E0432`.[0m
2023-10-12T13:08:18.6290347Z [RUSTC-TIMING] miri test:false 2.445
2023-10-12T13:08:18.6290720Z [0m[0m[1m[31merror[0m[1m:[0m could not compile `miri` (lib) due to previous error
2023-10-12T13:08:18.6290904Z 
```
This looks weird, like some build option wasn't passed correctly or something, as it cannot find 64-bit atomic types.

--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -39,6 +39,7 @@ use crate::core::config::flags;
 use crate::core::config::{DryRun, Target};
 use crate::core::config::{LlvmLibunwind, TargetSelection};
 use crate::utils::cache::{Interned, INTERNER};
+use crate::utils::exec::BootstrapCommand;
 use crate::utils::helpers::{
     self, dir_is_empty, exe, libdir, mtime, output, run, run_suppressed, symlink_dir,
     try_run_suppressed,
@@ -959,17 +960,32 @@ impl Build {
 
     /// Runs a command, printing out contextual info if it fails, and delaying errors until the build finishes.
     pub(crate) fn run_delaying_failure(&self, cmd: &mut Command) -> bool {
-        if !self.fail_fast {
-            #[allow(deprecated)] // can't use Build::try_run, that's us
-            if self.config.try_run(cmd).is_err() {
-                let mut failures = self.delayed_failures.borrow_mut();
-                failures.push(format!("{cmd:?}"));
-                return false;
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(cmd.delay_failure())
+    }
+
+    /// A centralized function for running commands that do not return output.
+    pub(crate) fn run_cmd<'a, C: Into<BootstrapCommand<'a>>>(&self, cmd: C) -> bool {
+        let command = cmd.into();
+        self.verbose(&format!("running: {command:?}"));
+
+        #[allow(deprecated)] // can't use Build::try_run, that's us
+        let result = self.config.try_run(command.command);
+
+        match result {
+            Ok(_) => true,
+            Err(_) => {
+                if command.delay_failure {
+                    let mut failures = self.delayed_failures.borrow_mut();
+                    failures.push(format!("{command:?}"));
+                    return false;
+                }
+                if self.fail_fast {
+                    exit!(1);
+                }
+                false
             }
-        } else {
-            self.run(cmd);
         }
-        true
     }
 
     pub fn is_verbose_than(&self, level: usize) -> bool {

--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -39,11 +39,8 @@ use crate::core::config::flags;
 use crate::core::config::{DryRun, Target};
 use crate::core::config::{LlvmLibunwind, TargetSelection};
 use crate::utils::cache::{Interned, INTERNER};
-use crate::utils::exec::BootstrapCommand;
-use crate::utils::helpers::{
-    self, dir_is_empty, exe, libdir, mtime, output, run, run_suppressed, symlink_dir,
-    try_run_suppressed,
-};
+use crate::utils::exec::{BehaviorOnFailure, BootstrapCommand};
+use crate::utils::helpers::{self, dir_is_empty, exe, libdir, mtime, output, symlink_dir};
 
 mod core;
 mod utils;
@@ -922,44 +919,30 @@ impl Build {
 
     /// Runs a command, printing out nice contextual information if it fails.
     fn run(&self, cmd: &mut Command) {
-        if self.config.dry_run() {
-            return;
-        }
-        self.verbose(&format!("running: {cmd:?}"));
-        run(cmd, self.is_verbose())
+        // FIXME: output mode -> status + err if self.is_verbose()
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(cmd.fail_fast());
     }
 
     /// Runs a command, printing out nice contextual information if it fails.
     fn run_quiet(&self, cmd: &mut Command) {
-        if self.config.dry_run() {
-            return;
-        }
-        self.verbose(&format!("running: {cmd:?}"));
-        run_suppressed(cmd)
+        // FIXME: output mode -> output + err
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(cmd.fail_fast());
     }
 
     /// Runs a command, printing out nice contextual information if it fails.
     /// Exits if the command failed to execute at all, otherwise returns its
     /// `status.success()`.
     fn run_quiet_delaying_failure(&self, cmd: &mut Command) -> bool {
-        if self.config.dry_run() {
-            return true;
-        }
-        if !self.fail_fast {
-            self.verbose(&format!("running: {cmd:?}"));
-            if !try_run_suppressed(cmd) {
-                let mut failures = self.delayed_failures.borrow_mut();
-                failures.push(format!("{cmd:?}"));
-                return false;
-            }
-        } else {
-            self.run_quiet(cmd);
-        }
-        true
+        // FIXME: output mode -> output + err
+        let cmd: BootstrapCommand<'_> = cmd.into();
+        self.run_cmd(cmd.delay_failure())
     }
 
     /// Runs a command, printing out contextual info if it fails, and delaying errors until the build finishes.
     pub(crate) fn run_delaying_failure(&self, cmd: &mut Command) -> bool {
+        // FIXME: output mode -> status + err if self.is_verbose()
         let cmd: BootstrapCommand<'_> = cmd.into();
         self.run_cmd(cmd.delay_failure())
     }
@@ -975,10 +958,16 @@ impl Build {
         match result {
             Ok(_) => true,
             Err(_) => {
-                if command.delay_failure {
-                    let mut failures = self.delayed_failures.borrow_mut();
-                    failures.push(format!("{command:?}"));
-                    return false;
+                if let Some(failure_behavior) = command.failure_behavior {
+                    match failure_behavior {
+                        BehaviorOnFailure::DelayFail => {
+                            let mut failures = self.delayed_failures.borrow_mut();
+                            failures.push(format!("{command:?}"));
+                        }
+                        BehaviorOnFailure::Exit => {
+                            exit!(1);
+                        }
+                    }
                 }
                 if self.fail_fast {
                     exit!(1);

--- a/src/bootstrap/src/lib.rs
+++ b/src/bootstrap/src/lib.rs
@@ -937,14 +937,18 @@ impl Build {
 
     /// Runs a command, printing out nice contextual information if it fails.
     fn run_quiet(&self, cmd: &mut Command) {
-        self.run_cmd(BootstrapCommand::from(cmd).fail_fast().output_mode(OutputMode::Suppress));
+        self.run_cmd(
+            BootstrapCommand::from(cmd).fail_fast().output_mode(OutputMode::SuppressOnSuccess),
+        );
     }
 
     /// Runs a command, printing out nice contextual information if it fails.
     /// Exits if the command failed to execute at all, otherwise returns its
     /// `status.success()`.
     fn run_quiet_delaying_failure(&self, cmd: &mut Command) -> bool {
-        self.run_cmd(BootstrapCommand::from(cmd).delay_failure().output_mode(OutputMode::Suppress))
+        self.run_cmd(
+            BootstrapCommand::from(cmd).delay_failure().output_mode(OutputMode::SuppressOnSuccess),
+        )
     }
 
     /// A centralized function for running commands that do not return output.
@@ -965,7 +969,7 @@ impl Build {
                 }),
                 matches!(mode, OutputMode::PrintAll),
             ),
-            OutputMode::Suppress => (command.command.output(), true),
+            OutputMode::SuppressOnSuccess => (command.command.output(), true),
         };
 
         let output = match output {

--- a/src/bootstrap/src/utils/exec.rs
+++ b/src/bootstrap/src/utils/exec.rs
@@ -1,0 +1,21 @@
+use std::process::Command;
+
+/// Wrapper around `std::process::Command`.
+#[derive(Debug)]
+pub struct BootstrapCommand<'a> {
+    pub command: &'a mut Command,
+    /// Report failure later instead of immediately.
+    pub delay_failure: bool,
+}
+
+impl<'a> BootstrapCommand<'a> {
+    pub fn delay_failure(self) -> Self {
+        Self { delay_failure: true, ..self }
+    }
+}
+
+impl<'a> From<&'a mut Command> for BootstrapCommand<'a> {
+    fn from(command: &'a mut Command) -> Self {
+        Self { command, delay_failure: false }
+    }
+}

--- a/src/bootstrap/src/utils/exec.rs
+++ b/src/bootstrap/src/utils/exec.rs
@@ -1,21 +1,32 @@
 use std::process::Command;
 
+/// What should be done when the command fails.
+#[derive(Debug, Copy, Clone)]
+pub enum BehaviorOnFailure {
+    /// Immediately stop bootstrap.
+    Exit,
+    /// Delay failure until the end of bootstrap invocation.
+    DelayFail,
+}
+
 /// Wrapper around `std::process::Command`.
 #[derive(Debug)]
 pub struct BootstrapCommand<'a> {
     pub command: &'a mut Command,
-    /// Report failure later instead of immediately.
-    pub delay_failure: bool,
+    pub failure_behavior: Option<BehaviorOnFailure>,
 }
 
 impl<'a> BootstrapCommand<'a> {
     pub fn delay_failure(self) -> Self {
-        Self { delay_failure: true, ..self }
+        Self { failure_behavior: Some(BehaviorOnFailure::DelayFail), ..self }
+    }
+    pub fn fail_fast(self) -> Self {
+        Self { failure_behavior: Some(BehaviorOnFailure::Exit), ..self }
     }
 }
 
 impl<'a> From<&'a mut Command> for BootstrapCommand<'a> {
     fn from(command: &'a mut Command) -> Self {
-        Self { command, delay_failure: false }
+        Self { command, failure_behavior: None }
     }
 }

--- a/src/bootstrap/src/utils/exec.rs
+++ b/src/bootstrap/src/utils/exec.rs
@@ -18,7 +18,7 @@ pub enum OutputMode {
     /// Print the output (by inheriting stdout/stderr).
     PrintOutput,
     /// Suppress the output if the command succeeds, otherwise print the output.
-    Suppress,
+    SuppressOnSuccess,
 }
 
 /// Wrapper around `std::process::Command`.
@@ -43,6 +43,6 @@ impl<'a> BootstrapCommand<'a> {
 
 impl<'a> From<&'a mut Command> for BootstrapCommand<'a> {
     fn from(command: &'a mut Command) -> Self {
-        Self { command, failure_behavior: None, output_mode: OutputMode::Suppress }
+        Self { command, failure_behavior: None, output_mode: OutputMode::SuppressOnSuccess }
     }
 }

--- a/src/bootstrap/src/utils/mod.rs
+++ b/src/bootstrap/src/utils/mod.rs
@@ -6,6 +6,7 @@ pub(crate) mod cache;
 pub(crate) mod cc_detect;
 pub(crate) mod channel;
 pub(crate) mod dylib;
+pub(crate) mod exec;
 pub(crate) mod helpers;
 pub(crate) mod job;
 #[cfg(feature = "build-metrics")]

